File: scripts/wordcloud.mjs
import { readFile, writeFile } from "fs/promises";
import cloud from "d3-cloud";
import { createCanvas } from "@napi-rs/canvas";

// Packs the /tags/ flex list into a spiral word cloud rendered as SVG anchors: real, crawlable,
// keyboard-focusable links, laid out deterministically at build time (no runtime JS, no layout shift).
const FILE = "public/tags/index.html";
const FONT = "sans-serif"; // measured and rendered with the same family so packing matches the glyphs
const MIN_PX = 16;
const MAX_PX = 68;
const PAD = 3;

const CONTAINER = /<div class=["']?tag-cloud["']? data-wordcloud>([\s\S]*?)<\/div>/;
const ANCHOR = /<a href=["']?([^"'\s>]+)["']?[^>]*?\sdata-w=["']?(\d+)["']?[^>]*>([\s\S]*?)<\/a>/g;

// small deterministic PRNG so the layout is byte-stable across builds
const makeRng = (seed) => () => {
  seed = (seed * 1664525 + 1013904223) >>> 0;
  return seed / 4294967296;
};

const html = await readFile(FILE, "utf-8");
const block = html.match(CONTAINER);
if (!block) {
  console.log("wordcloud: no tag-cloud container found, leaving flex fallback in place");
  process.exit(0);
}

const words = [];
for (const [, href, w, text] of block[1].matchAll(ANCHOR)) {
  words.push({ href, text: text.trim(), freq: Number(w) });
}
if (!words.length) {
  console.log("wordcloud: no tags parsed, leaving flex fallback in place");
  process.exit(0);
}

const weights = words.map((d) => Math.sqrt(d.freq + 1));
const wMin = Math.min(...weights);
const wSpan = Math.max(...weights) - wMin || 1;
words.forEach((d, i) => {
  d.t = (weights[i] - wMin) / wSpan;
  d.size = Math.round(MIN_PX + d.t * (MAX_PX - MIN_PX));
});

const glyphArea = words.reduce((a, d) => a + d.size * d.size * (d.text.length * 0.7 + 1), 0);

const layout = (width, height, seed) => {
  const rng = makeRng(seed);
  return new Promise((resolve) => {
    cloud()
      .size([width, height])
      .canvas(() => createCanvas(1, 1))
      .words(words.map((d) => ({ ...d })))
      .padding(PAD)
      .font(FONT)
      .fontSize((d) => d.size)
      .rotate(() => (rng() < 0.18 ? 90 : 0))
      .random(rng)
      .on("end", resolve)
      .start();
  });
};

// grow the canvas until every tag is placed (d3-cloud silently drops words that do not fit)
const aspect = 1.55;
let scale = 1.9;
let placed = [];
let width = 0;
let height = 0;
for (let attempt = 0; attempt < 7; attempt++) {
  const area = glyphArea * scale;
  width = Math.round(Math.sqrt(area * aspect));
  height = Math.round(width / aspect);
  placed = await layout(width, height, 0x9e3779b1);
  if (placed.length === words.length) break;
  scale *= 1.2;
}
if (placed.length < words.length) {
  console.log(`wordcloud: only ${placed.length}/${words.length} tags fit, keeping flex fallback`);
  process.exit(0);
}

// crop to the words' true extent so the cloud fills its width instead of floating in empty canvas
const mctx = createCanvas(10, 10).getContext("2d");
let x0 = Infinity;
let y0 = Infinity;
let x1 = -Infinity;
let y1 = -Infinity;
for (const d of placed) {
  mctx.font = `600 ${d.size}px ${FONT}`;
  const halfW = mctx.measureText(d.text).width / 2;
  const halfH = d.size * 0.62;
  const [ex, ey] = d.rotate ? [halfH, halfW] : [halfW, halfH];
  x0 = Math.min(x0, d.x - ex);
  y0 = Math.min(y0, d.y - ey);
  x1 = Math.max(x1, d.x + ex);
  y1 = Math.max(y1, d.y + ey);
}
const M = 6;
const vbX = (x0 - M).toFixed(1);
const vbY = (y0 - M).toFixed(1);
const vbW = (x1 - x0 + 2 * M).toFixed(1);
const vbH = (y1 - y0 + 2 * M).toFixed(1);

const esc = (s) => s.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;");
// screen-reader and tab order follow frequency, independent of the visual spiral placement
placed.sort((a, b) => b.freq - a.freq);

const anchors = placed
  .map((d) => {
    const cx = d.x.toFixed(1);
    const cy = d.y.toFixed(1);
    const rot = d.rotate ? ` transform="rotate(${d.rotate} ${cx} ${cy})"` : "";
    return (
      `<a href="${d.href}" aria-label="${esc(d.text)}">` +
      `<text x="${cx}" y="${cy}"${rot} font-size="${d.size}" style="--t:${d.t.toFixed(3)}">${esc(d.text)}</text>` +
      `</a>`
    );
  })
  .join("");

const svg =
  `<svg class="wordcloud" viewBox="${vbX} ${vbY} ${vbW} ${vbH}" role="group" ` +
  `aria-label="Blog topics, sized by how often each word appears" ` +
  `xmlns="http://www.w3.org/2000/svg" text-anchor="middle" dominant-baseline="central" font-family="${FONT}">` +
  `<title>Blog topics word cloud</title>${anchors}</svg>`;

// keep the flex list for narrow screens (readable), show the packed SVG where it has room; CSS swaps them
const replacement =
  `<div class="tag-cloud tag-cloud--flex">${block[1]}</div>` + `<div class="tag-cloud--svg">${svg}</div>`;
await writeFile(FILE, html.replace(CONTAINER, replacement));
console.log(`wordcloud: packed ${placed.length} tags into ${width}x${height} SVG`);
process.exit(0);

File: scripts/wordcloud.py
# /// script
# requires-python = ">=3.13"
# dependencies = ["pillow>=12.3", "numpy>=2.5.1", "fonttools[woff]>=4.63"]
# ///
"""Pack the /topics/ flex list into a horizontal spiral word cloud, emitted as SVG anchors
(crawlable, focusable links) at build time: no runtime JS, no layout shift.

Glyphs are measured with the exact font the page ships (static/fonts/tag-cloud.woff2), so the
browser paints the same widths the packer reserved and no two words overlap."""

# print is this CLI build script's progress output; stdout is intended
# ruff: noqa: T201

from __future__ import annotations

import io
import math
import re
from pathlib import Path
from typing import Final, NamedTuple

import numpy as np
from fontTools.ttLib import TTFont
from PIL import Image, ImageDraw, ImageFilter, ImageFont

_FILE: Final = Path("public/posts/index.html")
_WOFF2: Final = Path("static/fonts/tag-cloud.woff2")
_MIN_PX: Final = 16
_MAX_PX: Final = 68
_PAD: Final = 6  # clear pixels kept between neighbouring words
_MARGIN: Final = 6  # viewBox breathing room
_ASPECT: Final = 1.55
_CONTAINER: Final = re.compile(r'<div class=["\']?tag-cloud["\']? data-wordcloud>(.*?)</div>', re.DOTALL)
_ANCHOR: Final = re.compile(
    r'<a href=["\']?([^"\'\s>]+)["\']?[^>]*?\sdata-w=["\']?(\d+)["\']?[^>]*>(.*?)</a>', re.DOTALL
)


class Word(NamedTuple):
    href: str
    topic: str  # slug carried onto the SVG anchor so the filter JS can toggle it
    text: str
    freq: int
    size: int
    heat: float  # 0..1 position on the colour gradient


class Glyph(NamedTuple):
    tight: np.ndarray  # ink mask, for collision queries
    grown: np.ndarray  # ink dilated by _PAD, written into the occupancy grid
    left: int  # ink offset from the SVG anchor (horizontal centre, alphabetic baseline)
    top: int


class Placed(NamedTuple):
    word: Word
    x: float
    y: float
    left: float
    top: float
    right: float
    bottom: float


def main() -> None:
    html = _FILE.read_text()
    if not (block := _CONTAINER.search(html)):
        print("wordcloud: no tag-cloud container found, leaving flex fallback in place")
        return
    if not (words := parse_words(block.group(1))):
        print("wordcloud: no tags parsed, leaving flex fallback in place")
        return
    glyphs = build_glyphs(words)
    if not (placed := pack(words, glyphs)):
        print(f"wordcloud: could not fit {len(words)} tags, keeping flex fallback")
        return
    svg = render_svg(placed)
    replacement = (
        f'<div class="tag-cloud tag-cloud--flex">{block.group(1)}</div><div class="tag-cloud--svg">{svg}</div>'
    )
    _FILE.write_text(html[: block.start()] + replacement + html[block.end() :])
    print(f"wordcloud: packed {len(placed)} tags")


def parse_words(inner: str) -> list[Word]:
    raw = [(match[1], re.sub(r"\s+", " ", match[3]).strip(), int(match[2])) for match in _ANCHOR.finditer(inner)]
    if not raw:
        return []
    weights = [math.sqrt(freq + 1) for *_, freq in raw]
    lo = min(weights)
    span = (max(weights) - lo) or 1
    words = []
    for (href, text, freq), weight in zip(raw, weights, strict=True):
        heat = (weight - lo) / span
        topic = href.rsplit("topics=", 1)[-1]
        words.append(Word(href, topic, text, freq, round(_MIN_PX + heat * (_MAX_PX - _MIN_PX)), heat))
    return words


def build_glyphs(words: list[Word]) -> dict[Word, Glyph]:
    font = TTFont(str(_WOFF2))
    font.flavor = None
    buffer = io.BytesIO()
    font.save(buffer)
    ttf = buffer.getvalue()
    faces: dict[int, ImageFont.FreeTypeFont] = {}
    glyphs = {}
    for word in words:
        face = faces.setdefault(word.size, ImageFont.truetype(io.BytesIO(ttf), word.size))
        left, top, right, bottom = (round(edge) for edge in face.getbbox(word.text, anchor="ms"))
        image = Image.new("L", (right - left + 2 * _PAD, bottom - top + 2 * _PAD), 0)
        ImageDraw.Draw(image).text((_PAD - left, _PAD - top), word.text, font=face, fill=255, anchor="ms")
        tight = np.asarray(image) > 40
        grown = np.asarray(image.filter(ImageFilter.MaxFilter(2 * _PAD + 1))) > 40
        glyphs[word] = Glyph(tight, grown, left, top)
    return glyphs


def pack(words: list[Word], glyphs: dict[Word, Glyph]) -> list[Placed] | None:
    order = sorted(words, key=lambda word: (-word.size, word.text))
    ink = sum(int(glyph.tight.sum()) for glyph in glyphs.values())
    scale = 2.6
    for _ in range(7):
        width = round(math.sqrt(ink * scale * _ASPECT))
        if (placed := try_place(order, glyphs, width, round(width / _ASPECT))) is not None:
            return placed
        scale *= 1.3
    return None


def try_place(order: list[Word], glyphs: dict[Word, Glyph], width: int, height: int) -> list[Placed] | None:
    occ = np.zeros((height, width), dtype=bool)
    cx, cy = width // 2, height // 2
    placed = []
    for word in order:
        if (spot := spiral_place(occ, glyphs[word], cx, cy)) is None:
            return None
        dx, dy = spot
        glyph = glyphs[word]
        left, top = dx + glyph.left, dy + glyph.top
        placed.append(Placed(word, dx, dy, left, top, left + glyph.tight.shape[1], top + glyph.tight.shape[0]))
    return placed


def spiral_place(occ: np.ndarray, glyph: Glyph, cx: int, cy: int) -> tuple[float, float] | None:
    height, width = occ.shape
    th, tw = glyph.tight.shape
    limit = math.hypot(width, height)
    theta = 0.0
    while (radius := 4.0 * theta) < limit:
        dx = radius * math.cos(theta)
        dy = radius * math.sin(theta)
        gx = round(dx + glyph.left) + cx
        gy = round(dy + glyph.top) + cy
        in_bounds = gx - _PAD >= 0 and gy - _PAD >= 0 and gx + tw + _PAD <= width and gy + th + _PAD <= height
        if in_bounds and not occ[gy : gy + th, gx : gx + tw][glyph.tight].any():
            occ[gy - _PAD : gy - _PAD + glyph.grown.shape[0], gx - _PAD : gx - _PAD + glyph.grown.shape[1]] |= (
                glyph.grown
            )
            return dx, dy
        theta += 0.35 / (1.0 + 0.02 * theta)
    return None


def render_svg(placed: list[Placed]) -> str:
    x0 = min(word.left for word in placed) - _MARGIN
    y0 = min(word.top for word in placed) - _MARGIN
    x1 = max(word.right for word in placed) + _MARGIN
    y1 = max(word.bottom for word in placed) + _MARGIN
    body = "".join(
        f'<a href="{word.word.href}" data-topic="{esc(word.word.topic)}" aria-label="{esc(word.word.text)}">'
        f'<text x="{word.x:.1f}" y="{word.y:.1f}" font-size="{word.word.size}" '
        f'style="--t:{word.word.heat:.3f}">{esc(word.word.text)}</text>'
        f"</a>"
        # screen-reader and tab order follow frequency, independent of the visual spiral placement
        for word in sorted(placed, key=lambda word: -word.word.freq)
    )
    return (
        f'<svg class="wordcloud" viewBox="{x0:.1f} {y0:.1f} {x1 - x0:.1f} {y1 - y0:.1f}" role="group" '
        f'aria-label="Blog topics, sized by how often each word appears" '
        f'xmlns="http://www.w3.org/2000/svg" text-anchor="middle" font-family="TagCloud">'
        f"<title>Blog topics word cloud</title>{body}</svg>"
    )


def esc(text: str) -> str:
    return text.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;")


if __name__ == "__main__":
    main()
